### PR TITLE
NIAD-1951: Add ObservationCommentMapper to allow mapping NarrativeSta…

### PR DESCRIPTION
**Pull request by mrogowska**
…tement to Observation resource

### CODE REVIEW

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java [hunk 1]
```diff
@@ -108,8 +116,13 @@ private Patient mapPatient(RCMRMT030101UK04EhrExtract ehrExtract, Organization o
             .toList();
     }
 
+    private List<Observation> mapObservationComments(RCMRMT030101UK04EhrExtract ehrExtract, Patient patient, List<Encounter> encounters) {
+        return observationCommentMapper.mapObservations(ehrExtract, patient, encounters);
+    }
+
     private List<Observation> mapObservations(RCMRMT030101UK04EhrExtract ehrExtract, Patient patient, List<Encounter> encounters) {
         return observationMapper.mapObservations(ehrExtract, patient, encounters);
+
```
Comment (rorysproule1) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java:125]: Remove extra line pls

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    private List<RCMRMT030101UK04EhrComposition> getCompositionsWithNarrativeStatement(RCMRMT030101UK04EhrExtract ehrExtract) {
+        return ehrExtract.getComponent()
+            .stream()
+            .flatMap(component -> component.getEhrFolder().getComponent().stream())
+            .map(RCMRMT030101UK04Component3::getEhrComposition)
+            .filter(ehrComposition -> ehrComposition.getComponent()
+                .stream()
+                .map(RCMRMT030101UK04Component4::getNarrativeStatement)
+                .anyMatch(Objects::nonNull))
```
Comment (rorysproule1) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java:145]: Remember you need to filter out any NarrativeStatements that contain an element with a referredToExternalDocument element eg.

`<reference typeCode="REFR">
<referredToExternalDocument classCode="DOC" moodCode="EVN">
                <id root="394559384658936"/>
                <code code="37251000000104" codeSystem="2.16.840.1.113883.2.1.3.2.4.15" displayName="Other digital signal"/>
                <text mediaType="text/plain">
                    <reference value="file://localhost/AbsentAttachment394559384658936.txt"/>
                </text>
            </referredToExternalDocument>
        </reference>`

**Review by rorysproule1 — CHANGES_REQUESTED** [review]

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            .stream()
+            .flatMap(composition -> composition.getComponent().stream())
+            .map(RCMRMT030101UK04Component4::getNarrativeStatement)
+            .filter(Objects::nonNull)
+            .map(narrativeStatement -> {
+                var narrativeStatementId = narrativeStatement.getId();
+                var observation = new Observation();
+                observation.setId(narrativeStatement.getId().getRoot());
+                observation.setMeta(createMeta());
+                observation.setStatus(Observation.ObservationStatus.FINAL);
```
[the remark this answers is not in the document]
Comment (boydstep) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java:54]: Sure thing

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    private RCMRMT030101UK04NarrativeStatement getNarrativeStatement(RCMRMT030101UK04EhrExtract ehrExtract) {
+        return ehrExtract.getComponent().get(0)
+            .getEhrFolder().getComponent().get(0)
+            .getEhrComposition().getComponent().get(0)
+            .getNarrativeStatement();
+    }
+
+    @SneakyThrows
+    private RCMRMT030101UK04EhrExtract unmarshallCodeElement(String fileName) {
```
Comment (mrogowska) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapperTest.java:138]: `unmarshallEhrExtractElement` or `unmarshallEhrExtract` could be more suitable name :D

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+    private CodeableConcept createCodeableConcept() {
+        var codeableConcept = new CodeableConcept();
+        codeableConcept.setCoding(
+            Collections.singletonList(new Coding(CODING_SYSTEM, CODING_CODE, CODING_DISPLAY)));
+
+        return codeableConcept;
+    }
+
+    private List<RCMRMT030101UK04EhrComposition> getCompositionsWithNarrativeStatement(RCMRMT030101UK04EhrExtract ehrExtract) {
```
Comment (mrogowska) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java:137]: Could this method return narrative statements? Then you wouldn't have to do the filtering in the `mapObservations` method again

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapperTest.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    @Test
+    public void mapObservationWithFullDataNoMappedEncounter() {
+        var ehrExtract = unmarshallCodeElement("single_narrative_statement.xml");
+
+        List<Observation> observations =
+            observationCommentMapper.mapObservations(ehrExtract, patient, Collections.emptyList());
+
+        // Calling `getContext` auto creates a Reference object so asserting the reference is null
```
Comment (bbell-kainos) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapperTest.java:116]: I've seen comments used throughout the PR, are they entirely necessary? I guess its up for debate, although my argument would be that comments (at present) add noise and are usually only used in this project to highlight important implementations yet to be developed (TODOs). In the case of these tests, maybe rename the tests themselves to note what the comment says?

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java (hunk 1 above)
Comment (boydstep) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/service/BundleMapperService.java]: Removed

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java (hunk 2 above)
Comment (boydstep) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java]: Ah yes, thank you for spotting this.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapperTest.java (hunk 4 above)
Comment (boydstep) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapperTest.java]: Yes, thanks for spotting that

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java (hunk 5 above)
Comment (boydstep) [reply gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapper.java]: Yup!

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapperTest.java (hunk 6 above)
Comment (boydstep) [reply gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ObservationCommentMapperTest.java:125]: As discussed privately, we've ditched one and left the other two

**Review by rorysproule1 — DISMISSED** [review]

**Review by bbell-kainos — APPROVED** [review]

**Review by rorysproule1 — APPROVED** [review]

**Pull request merged**